fetch/fetch.go

```diff
@@ -479,26 +507,115 @@ func (f *Fetch) requestHashBatchFromPeers() {
 	}
 	f.activeReqM.Unlock()
 
-	// send in batches
-	for i := 0; i < len(requestList); i += f.cfg.BatchSize {
-		j := i + f.cfg.BatchSize
-		if j > len(requestList) {
-			j = len(requestList)
+	f.send(requestList)
+}
+
+func (f *Fetch) send(requests []requestMessage) {
+	if len(requests) == 0 {
+		return
+	}
+	if f.stopped() {
+		return
+	}
+
+	retries := 0
+	for {
+		peer2batches := f.organizeRequests(requests)
+		requests = requests[:0]
+
+		for peer, peerBatches := range peer2batches {
+			for _, reqs := range peerBatches {
+				err := f.sendBatch(peer, reqs)
+				if err != nil {
+					requests = append(requests, reqs...)
+				}
+			}
+		}
+
+		if len(requests) == 0 {
+			break
 		}
-		f.sendBatch(requestList[i:j])
+
+		retries++
+		if retries > f.cfg.MaxRetriesForBatch {
```

Comment:
why should we need to add another level of retries?
per peer and per request seems adequate already. whenever a request fails, it'll be back to the active queue and go through this again anyway. it's set to 100 times now, but i am hoping with you change we can reduce it to something like 5.

Reply:
hmm, at that moment I was sure it was needed :) removed.